User windows fix

mhackethal opened the pull request:
<!--- Provide a short summary of your changes in the Title above -->

Provide expected attributes in user resource. Missing usage of cache for windows accounts
reflect windows attributes to be also available for linux.

## Description
It solves: #4432 and parts of #4174 
Expected behaviour on linux and windows should be equal.

## Related Issue
<!--- If you are suggesting a new feature or change, please create an issue first -->
Fixes #4432
#4174 
<!--- Please link to the issue, discourse, or stackoverflow here: -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New content (non-breaking change)
- [ ] Breaking change (a content change which would break existing functionality or processes)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the **CONTRIBUTING** document.

mhackethal committed:
add cache credentials/meta_info to windows user resource

mhackethal committed:
Update users.rb to include new attributes

mhackethal committed:
Update user_test.rb

miah commented on lib/inspec/resources/users.rb:
```diff
@@ -425,10 +463,26 @@ def credentials(username)
         multiple_values: false
       ).params
 
+      cmd = inspec.command("echo $((($(date +%s) - $(date -d '#{params["Last password change"]}' '+%s'))/86400))")
```
Do we need to do math in a `inspec.command`?

mhackethal replied:
hi miah, i just pulled this request because no one answered to my "discussion" Topic :)
#4432.
I opened a bug and described why i did math calculation on the target. The reason for doing it on the target is the localization. Each target could have a different localization and therefor ist close to impossible to calculate it by ruby. Next to this the ruby "date" class is not available from inside inspec. Same issue as for some AWS ressources.

I am totally open minded at this Topic. Any discussion and ideas are welcome and should be documented issue #4432.

zenspider replied:
"localization" is the reason why you subtract and divide by 86400 on the client?

mhackethal replied:
There are 3 reason for calculation. 
1) Localization of the client 
2) harmonize output to windows. 
3) Ruby date class not a part of "inspec" package.

clintoncwolfe replied:
Hi @mhackethal ,
  [`date`](https://ruby-doc.org/stdlib-2.6.4/libdoc/date/rdoc/index.html), is the standard Ruby library that includes the `Date` class, and as such, it is included with every Ruby installation. You can (and should) do any conversion in Ruby.

zenspider replied:
I've got to push back on this one.

1. Too complicated.
2. `date -d` doesn't do what you think it does (on all platforms).
3.  Ruby can do all the conversions for you:

```ruby
>> require "date"
=> true
>> d0 = Date.parse "2019-01-01"
=> #<Date: 2019-01-01 ((2458485j,0s,0n),+0s,2299161j)>
>> Date.today - d0
=> (266/1)
```

mhackethal replied:
i am sorry. Youre totally right. Looks like i did something wrong on my end. (big sorry)
On my first tests i received an error when i used require "date" so i thought its not usable.
But looks like i did something wrong - its working now.
```

      #cmd = inspec.command("echo $((($(date +%s) - $(date -d '#{params["Last password change"]}' '+%s'))/86400))")
      #dayslastset = convert_to_i(cmd.stdout.chomp) if cmd.exit_status == 0
      dparse = Date.parse "#{params['Last password change']}"
      dayslastset = (Date.today - dparse).to_i
```
I changed the the code and tested on:
RHEL 6 , 7
Ubuntu 12.x,14,x,16,x
Debian 9.x
Debian 10.x

```
You are currently running on:
    Name:      redhat
    Families:  redhat, linux, unix, os
    Release:   6.10
    Arch:      x86_64
inspec> describe user("root") do
inspec>   its('passwordage') { should eq 0 }
inspec> end

Profile: inspec-shell
Version: (not specified)
  User root
     [FAIL]  passwordage should eq 0
     expected: 0
          got: 9
     (compared using ==)
Test Summary: 0 successful, 1 failure, 0 skipped
```
```
You are currently running on:
    Name:      ubuntu
    Families:  debian, linux, unix, os
    Release:   16.04
    Arch:      x86_64
inspec> describe user("root") do
inspec>   its('passwordage') { should eq 0 }
inspec> end
Profile: inspec-shell
Version: (not specified)
  User root
     [FAIL]  passwordage should eq 0
     expected: 0
          got: 76
     (compared using ==)
Test Summary: 0 successful, 1 failure, 0 skipped
```

I tested both ways onetime with the "cmd" parser and one time with the ruby parser. On my systems the result was equal on both.
I personally would prefer the "cmd" way because i have the following situation in my mind:
- InSpec uses : dd/mm/yyyy
- Target uses : mm/dd/yyyy
Same for yyyy/mm/dd vs. yyyy/dd/mm

Any hint on which distro "date -d" will not do what i expect it to do? (just for personal intrest on this topic).

I will push an update to the code later. Thank you for your time and supporting me on this topic!

mhackethal replied:
All done and updated.
Does anyone can give me some suggestion about CodeClimat and the failing BK Test for Windows?
Is this something i should solve?

kekaichinose replied:
@miah any feedback for @mhackethal on this comment?

miah replied:
I've reviewed the codeclimate complaints here and they are valid, but don't need to be addressed by this PR.

mhackethal replied:
Can i consider this as no More Tasks for me?

miah replied:
Depends on the other reviewers =)

miah commented on lib/inspec/resources/users.rb:
```diff
@@ -425,10 +463,26 @@ def credentials(username)
+        dayslastset = convert_to_i(cmd.stdout.chomp)
+      end
+
+      cmd = inspec.command("lastb -w -a | awk '{print $1}' | grep #{username} | wc -l")
```
I think we could just do `lastb -w -a #{username}|wc -l`

But also, we'll need to subtract 2 from that number. As we'll have a empty line and informational line about btmp.

```
$ sudo lastb -w -a miah
miah     ssh:notty    Fri Sep  6 11:12 - 11:12  (00:00)     ::1
miah     ssh:notty    Fri Sep  6 11:12 - 11:12  (00:00)     ::1
miah     ssh:notty    Fri Sep  6 11:12 - 11:12  (00:00)     ::1

btmp begins Fri Sep  6 11:12:07 2019
```

We'll also need to note that this is not a cumulative number of all bad login attempts from that user. Its only since the creation of that btmp file; because this stores all bad login attempts systems on the public internet might refresh this more often. Some distro's might even recreate this file on reboot. I don't have all the details on the lifecycle of that file, but it might be good to consider.

mhackethal replied:
Same as your previous comment -> discussion should be done on issue #4432. I noted that specifics down but there is now way to determinate an all time bad Login Count. Because of the log Rotation and Retention Policy. This Count is Always depending on the local System Setup. (Same like on Windows). But every Administrator should be Aware of this. New ideas or any comment is welcome.

mhackethal replied:
Just some additional remarks. I think its better to keep go with "awk&grep" on your System i have to calculate "-2" but Maybe on another distro there are 3 empty lines are an additional info line. This could break the calculation. By using awk and grep i can ensure that it only Counts the lines of badloginattempts. The Performance should have no effect in that case.

zenspider replied:
I don't see any need for the `awk` part. I'd go for grep + wc or even `grep -c` (but I'm not sure how universal the count flag is).

zenspider replied:
And discussion about code should go next to the code so we know what we're talking about.

miah reviewed (changes requested):
Thanks for the contribution @mhackethal =)

I would love to add this functionality, thanks for starting the work on it!

We need tests and documentation before we can merge this work in. If you have any questions please don't hesitate to ask.

zenspider commented on lib/inspec/resources/users.rb:
```diff
@@ -425,10 +463,26 @@ def credentials(username)
+        badpasswordattempts = nil
+      else
+        badpasswordattempts = convert_to_i(cmd.stdout.chomp)
+      end
```
```ruby
# lift assignment
badpasswordattempts = if cmd.exit_status != 0
                        nil
                      else
                        convert_to_i(cmd.stdout.chomp)
                      end

# inline conditional
badpasswordattempts = cmd.exit_status != 0 ? nil :convert_to_i(cmd.stdout.chomp)

# simplify
badpasswordattempts = convert_to_i(cmd.stdout.chomp) if cmd.exit_status == 0
```

mhackethal replied:
Thanks for this ruby expert tips :) 
i will do the requested updates.

zenspider commented on lib/inspec/resources/users.rb:
```diff
@@ -573,6 +629,31 @@ def identity(username)
+        minpasswordlength: res[:minpasswordlength],
+        passwordage: res[:passwordage],
+      }
+    end
```
These new methods need tests.

mhackethal replied:
Ill check if can find where to put that. Again thanks!

mhackethal committed:
Update users.rb

Signed-off-by: mhackethal <mh@it31.de>

mhackethal committed:
remove awk on lastb

mhackethal committed:
first test. Signed-off-by: mhackethal  <mh@it31.de>

mhackethal committed:
fix whitespace. Signed-off-by: mhackethal  <mh@it31.de>

mhackethal committed:
Update user_test.rb

adjust expectation of methods to be "nil" on windows

mhackethal committed:
adjust user_test.rb

mhackethal committed:
update user_test.rb
Signed-off-by: mhackethal <mh@it31.de>

mhackethal committed:
Update documentation for user(s) resource. Signed-off-by: mhackethal  <mh@it31.de>

mhackethal committed:
Ruby calculation for passwordage. 

Change calculation for passwordage from inspec.cmd to ruby date calculation.
Added require "date"
Parse provided date and calculate against Date.today.
Signed-off-by: mhackethal <mh@it31.de>

mhackethal committed:
fix whitespace. Signed-off-by: mhackethal <mh@it31.de>

miah reviewed (approved):
Thanks @mhackethal!
